examples/tide/main.rs

```diff
@@ -0,0 +1,186 @@
+
+extern crate webauthn_rs;
+
+use askama::Template;
+
+use std::sync::Arc;
+
+use rand::prelude::*;
+
+use webauthn_rs::ephemeral::WebauthnEphemeralConfig;
+use webauthn_rs::proto::{PublicKeyCredential, RegisterPublicKeyCredential};
+
+mod actors;
+
+use crate::actors::*;
+
+#[derive(Template)]
+#[template(path = "index.html")]
+struct IndexTemplate;
+
+type AppState = Arc<WebauthnActor>;
+
+#[derive(Debug, StructOpt)]
+struct CmdOptions {
+    #[structopt(short = "d", long = "debug")]
+    debug: bool,
+    #[structopt(short = "p", long = "prefix", default_value = "/auth")]
+    prefix: String,
+    #[structopt(short = "n", long = "name", default_value = "localhost")]
+    rp_name: String,
+    #[structopt(short = "o", long = "origin", default_value = "http://localhost:8080")]
+    /// Must match your sites domain/port/url
+    rp_origin: String,
+    #[structopt(short = "i", long = "id", default_value = "localhost")]
+    rp_id: String,
+    #[structopt(short = "b", long = "bind", default_value = "localhost:8080")]
+    bind: String,
+    #[structopt(short = "s", long = "tls")]
+    enable_tls: bool,
+}
+
+async fn index_view(mut request: tide::Request<AppState>) -> tide::Result {
+    let session = request.session_mut();
+    let some_userid = session.get::<String>("userid");
+
+    if some_userid.is_none() {
+        if session.insert("anonymous", true).is_err() {
+            return Ok(tide::Response::new(tide::StatusCode::InternalServerError));
+        }
+    }
+
+    Ok(askama_tide::into_response(&IndexTemplate {}, "html"))
+}
+
+async fn challenge_register(request: tide::Request<AppState>) -> tide::Result {
+    let username = request.param("username")?.parse()?;
+    let actor_res = request.state().challenge_register(username).await;
+    let res = match actor_res {
+        Ok(chal) => tide::Response::builder(tide::StatusCode::Ok)
+            .body(tide::Body::from_json(&chal)?)
+            .build(),
+        Err(e) => {
+            tide::log::debug!("challenge_register -> {:?}", e);
+            tide::Response::new(tide::StatusCode::InternalServerError)
+        }
+    };
+    Ok(res)
+}
+
+async fn challenge_login(request: tide::Request<AppState>) -> tide::Result {
+    let username = request.param("username")?.parse()?;
+    let actor_res = request.state().challenge_authenticate(&username).await;
+    let res = match actor_res {
+        Ok(chal) => tide::Response::builder(tide::StatusCode::Ok)
+            .body(tide::Body::from_json(&chal)?)
+            .build(),
+        Err(e) => {
+            tide::log::debug!("challenge_login -> {:?}", e);
+            tide::Response::new(tide::StatusCode::InternalServerError)
+        }
+    };
+    Ok(res)
+}
+
+async fn register(mut request: tide::Request<AppState>) -> tide::Result {
+    let username = request.param("username")?.parse()?;
+    let reg = request.body_json::<RegisterPublicKeyCredential>().await?;
+    let actor_res = request.state().register(&username, &reg).await;
```

Comment:
We could kick these futures off with `async_std::task::spawn` but everything in tide ultimately gets run in the async-std runtime anyway so I'm not sure it makes a difference.

Reply:
I don't think it matters to much in this case, I think maybe task::spawn is over-kill. But as mentioned we need async mutex

Reply:
Fixed